aiida_vasp/workflows/relax.py

```diff
@@ -0,0 +1,298 @@
+        if self.inputs.relax.shape.value:
+            incar.encut = encut.value
+
+        if 'incar_add' in self.inputs:
+            try:
+                incar_overrides = clean_incar_overrides(self.inputs)
+                incar.update(incar_overrides)
+            except ValueError as err:
+                self._fail_compat(exception=err)
+
+        for param in [ibrion, isif, encut]:
+            errors, warnings = param.clean(incar)
+            for error in errors:
+                self.logger.error(error)
+            for warning in warnings:
+                self.report(warning)
+            if errors:
+                self._fail_compat(exception=ValueError(errors[0]))
+
+        return incar
+
+    def _clean_kpoints(self):
+        """Use or create a kpoints mesh depending on user input."""
+        kpoints = None
+        if 'mesh' in self.inputs.kpoints:
+            kpoints = self.inputs.kpoints.mesh
+        elif 'distance' in self.inputs.kpoints:
+            kpoints = get_data_node(
+                'array.kpoints', cell_from_structure=self.inputs.structure, kpoints_mesh_from_density=self.inputs.kpoints.distance.value)
+        else:
+            self._fail_compat(exception=ValueError('either kpoints.mesh or kpoints.distance is required'))
+        return kpoints
+
+    def setup(self):
+        """Store exposed inputs in the context."""
+        self.ctx.current_structure = self.inputs.structure
+        self.ctx.current_restart_folder = None
+        self.ctx.is_converged = False
+        self.ctx.iteration = 0
+        self.ctx.workchains = []
+
+        self.ctx.inputs = AttributeDict()
+        self.ctx.inputs.code = self.inputs.code
+        self.ctx.inputs.structure = self.inputs.structure
+        self.ctx.inputs.potcar_family = self.inputs.potcar_family
+        self.ctx.inputs.potcar_mapping = self.inputs.potcar_mapping
+        self.ctx.inputs.options = self.inputs.options
+        if 'max_iterations' in self.inputs.restart:
+            self.ctx.inputs.max_iterations = self.inputs.restart.max_iterations
+
+    def validate_inputs(self):
+        self.ctx.inputs.kpoints = self._clean_kpoints()
+        self.ctx.inputs.incar = self._clean_incar(RELAXATION_INCAR_TEMPLATE)
+
+    def should_relax(self):
+        within_max_iterations = bool(self.ctx.iteration < self.inputs.convergence.max_iterations.value)
+        return bool(within_max_iterations and not self.ctx.is_converged)
+
+    def run_relax(self):
+        """Run the BaseVaspWf for the relaxation."""
+        if self.ctx.current_restart_folder:
+            self.ctx.inputs.restart_folder = self.ctx.current_restart_folder
+        inputs = prepare_process_inputs(self.ctx.inputs)
+        running = self.submit(VaspBaseWf, **inputs)
+        self.ctx.iteration += 1
+
+        self.report('launching VaspBaseWf{}'.format(running))
+
+        return ToContext(workchains=append_(running))
+
+    def inspect_relax(self):
```

Comment:
This lacks incorporation of the most important and most used relaxation parameter, the EDIFFG. Usually negative numbers are used and then it is the forces. If one is posing structure properties in the convergence loop one should check if VASP has converged to the specified EDIFFG. As such there is no point of restarting. If not, either NSW was set to low and one can just restart X number of times, which basically is the same as bumping NSW up X times. If not, some action need to be taken. The most useful relaxation workflow for most users would do an initial IBRION=2 followed by an IBRION=1. This saves quite a lot of computational resources and is typically the workflow of a more experienced VASP user. However, one could argue that this can reside in a top relax workchain, in addition to the convergence test on the structure. The convergence criteria for the structure is in fact not very often used. One often accepts the structure as is, given that the forces are low internally (looking for stable structures and the nature tends to reduce forces). There is no reason not to have it, since you have already added it, but the EDIFFG should enter, potentially also iterations on IBRION if we do not want to add another relaxation workchain.

Reply:
I haven't gotten around to implementing this yet. Would you agree to release the workchain into develop without changing IBRION in the iterations and adding that functionality in a later PR?

Reply:
Sure, let us do that.